lib/portlayer/exec/exec.go

```diff
@@ -171,6 +195,144 @@ func eventCallback(ie events.Event) {
+						}
+					}
+				case types.VirtualMachinePowerState:
+					if v != types.VirtualMachinePowerStatePoweredOn {
+						// Give up if the vm has powered off
+						poweredoff = true
+						return true
+					}
+				}
+
+			}
+			return false
+		}
+		log.Debugf("Waiting for container(%s) key %s", container.ExecConfig.ID, k)
+
+		err := container.vm.WaitForExtraConfig(ctx, waitFunc)
+		if poweredoff {
+			log.Errorf("Failed to wait container started for vm is powered off")
+			return false
+		}
+		if err != nil {
+			log.Errorf("Failed to wait container started: %s", err)
+		}
+	}
+	return true
+}
+
+func setStopTime(ctx context.Context, id string) error {
+	operation := func() error {
+		log.Debugf("Updating container(%s) stopTime", id)
+		h := GetContainer(ctx, uid.Parse(id))
+		if h == nil {
+			log.Errorf("Failed to get handle of container(%s), stop retry", id)
+			return nil
+		}
+		defer h.Close()
+
+		if h.vm == nil {
+			log.Errorf("Container(%s) update failed for vm is not found", h.ExecConfig.ID)
+			return nil
+		}
+
+		var update bool
+		if h.ExecConfig == nil {
+			return &retryError{fmt.Errorf("Failed to get container(%s) configuration, retry later", h.ExecConfig.ID)}
+		}
+		if h.Runtime.PowerState != types.VirtualMachinePowerStatePoweredOff {
+			log.Debugf("container(%s) power state is changed to %s, stop configuration updating for powered off event", h.ExecConfig.ID, h.Runtime.PowerState)
+			return nil
+		}
+		for _, sc := range h.ExecConfig.Sessions {
+			if sc.StopTime == 0 {
+				sc.StopTime = time.Now().UTC().Unix()
+				sc.StartTime = 0
+				log.Debugf("Container(%s) session %s stop time is set to %s", h.ExecConfig.ID, sc.ID, time.Unix(sc.StopTime, 0))
+				update = true
+			} else {
+				log.Debugf("Container(%s) session %s stop time is already set to %s", h.ExecConfig.ID, sc.ID, time.Unix(sc.StopTime, 0))
+			}
+		}
+		if !update {
+			log.Debugf("No need to update container(%s) configuration", h.ExecConfig.ID)
+			return nil
+		}
+		log.Debugf("commit container(%s)", h.ExecConfig.ID)
+		// set waittime to 0 for vm is already stopped, no need to wait again
+		return h.Commit(ctx, h.vm.Session, new(int32))
+	}
+	retryOnError := func(err error) bool {
+		switch err.(type) {
+		case ConcurrentAccessError, retryError:
+			log.Debugf("Retry container commit")
+			return true
+		default:
+			return false
+		}
+	}
+	return backoff.Retry(operation, retryOnError)
+}
+
+func containerTimeCallback(ie events.Event) {
+	container := Containers.Container(ie.Reference())
+	if container == nil {
+		log.Debugf("Container(%s) is not found", ie.Reference())
+	}
+	go func() {
+		switch ie.String() {
+		case events.ContainerPoweredOn:
+			// container powered on event received, still need to wait container startTime set by tether, to sync between portlayer and vmx
+			// we'll do this in a go routine to avoid blocking
+			if !waitForContainerStarted(container) {
+				return
+			}
+			fallthrough
```

Comment:
This is the intentionally fall through. We need to refresh after the vm is powered on OOB, and then wait for at most 3 minutes. If we got container starttime set, that means the container is started.
Which is similar to docker start, which will wait for Started instead of starttime there, because docker start reset container started attribute before poweron VM. I need to use starttime for OOB poweron, here.
After that, should sync up with vsphere, for tether already configuration.

And we cannot skip the refresh immediately after powered on event received, because at that time, there are some attributes set, and if in any chance we failed to wait container starttime, we'll not refresh again.